src/components/layerpublisher/LayerPublisher.js

```diff
@@ -24,43 +43,56 @@ function info(fileName) {
   PubSub.publish(STATUS_INFO, `Loaded ${fileName}.`);
 }
 
+function fetchDataFromDZI(layerType, data) {
+  return fetch(data[layerType].tileSource).then(response => response.text())
+    .then(str => (new window.DOMParser()).parseFromString(str, 'text/xml'))
+    .then((layer) => {
+      const { tileSource } = data[layerType];
+      data[layerType].tileSource = tileSource.substring(0, tileSource.lastIndexOf('/')); // eslint-disable-line no-param-reassign
+      // eslint-disable-next-line no-param-reassign
+      data[layerType].height = layer.getElementsByTagName('Size')[0].attributes.Height.value;
+      // eslint-disable-next-line no-param-reassign
+      data[layerType].width = layer.getElementsByTagName('Size')[0].attributes.Width.value;
+      // eslint-disable-next-line no-param-reassign
+      data[layerType].tileSize = layer.getElementsByTagName('Image')[0].attributes.TileSize.value;
+      return data;
+    });
+}
+
+function fetchImageMetadata(data) {
+  return fetchDataFromDZI('nuclei', data).then(res => fetchDataFromDZI('polyT', res));
+}
+
+function publishLayer(data, type, name, url) {
+  const schema = typeToSchema[type];
+  if (!schema) {
+    throw Error(`No schema for ${type}`);
+  }
+  const validate = new Ajv().compile(schema);
+  const valid = validate(data);
+  if (valid) {
+    if (type === 'IMAGES') {
+      fetchImageMetadata(data).then((resData) => {
+        PubSub.publish(typeToEvent[type], resData);
+        info(name);
+      });
+    } else {
+      PubSub.publish(typeToEvent[type], data);
+      info(name);
+    }
+  } else {
+    const failureReason = JSON.stringify(validate.errors, null, 2);
+    warn(`Error while validating ${name}. Details in console.`);
+    console.warn(`"${name}" (${type}) from ${url}: validation failed`, failureReason);
+  }
```

Comment:
Style comments:
- Consider a factoring `PubSub.publish(typeToEvent[type], data); info(name);` into a function.
- Consider rearranging as a guard clause:
```
  if (!valid) {
    const failureReason = JSON.stringify(validate.errors, null, 2);
    warn(`Error while validating ${name}. Details in console.`);
    console.warn(`"${name}" (${type}) from ${url}: validation failed`, failureReason);
    return;
  }
  if (type === 'IMAGES') {
    ...
  } else {
    ...
  }
```
Less if nesting / shorter clause first makes it easier to read.

Reply:
I'll do the guard clause, and I can do the function - when I do the function, doing that would validate what I commented about moving the `typeToSchema` and `typeToEvent` out into their own configuration

Reply:
cool, good point.